File: mod-common/src/main/java/com/dreamdisplays/net/ReportPacket.java
package com.dreamdisplays.net;

import com.dreamdisplays.PlatformlessInitializer;

import java.util.UUID;
import net.minecraft.core.UUIDUtil;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;

// Packet for reporting a display
public record ReportPacket(UUID id) implements CustomPacketPayload {
    public static final Type<ReportPacket> PACKET_ID =
            new Type<>(Identifier.fromNamespaceAndPath(PlatformlessInitializer.MOD_ID, "report"));

    public static final StreamCodec<FriendlyByteBuf, ReportPacket> PACKET_CODEC =
            StreamCodec.of(
                    (buf, packet) -> UUIDUtil.STREAM_CODEC.encode(buf, packet.id()),
                    (buf) -> {
                        UUID id = UUIDUtil.STREAM_CODEC.decode(buf);
                        return new ReportPacket(id);
                    });

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return PACKET_ID;
    }
}

File: common/src/main/kotlin/com/dreamdisplays/net/c2s/ReportPacket.kt
package com.dreamdisplays.net.c2s

import com.dreamdisplays.net.common.helpers.createType
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamCodec.of
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.network.protocol.common.custom.CustomPacketPayload.Type
import java.util.*

data class ReportPacket(val uuid: UUID) : CustomPacketPayload {
    override fun type(): Type<out CustomPacketPayload> = PACKET_ID

    companion object {
        val PACKET_ID: Type<ReportPacket> = createType("report")
        val PACKET_CODEC: StreamCodec<FriendlyByteBuf, ReportPacket> = of(
            { buf, packet -> buf.writeUUID(packet.uuid) },
            { buf -> ReportPacket(buf.readUUID()) }
        )
    }
}
